src/Foundatio.Aliyun/Storage/AliyunFileStorage.cs

```diff
@@ -0,0 +1,199 @@
+                return response.StatusCode == HttpStatusCode.NotFound;
+            }
+            return false;
+        }
+
+        private string NormalizePath(string path) {
+            return path?.Replace('\\', '/');
+        }
+
+        public async Task<Stream> GetFileStreamAsync(string path, CancellationToken cancellationToken = default(CancellationToken)) {
+            if (string.IsNullOrEmpty(path)) {
+                throw new ArgumentNullException(nameof(path));
+            }
+            var response = await Task.Factory.FromAsync(
+                (request, callback, state) => _client.BeginGetObject(request, callback, state),
+                result => _client.EndGetObject(result),
+                new GetObjectRequest(_bucketName, NormalizePath(path)), null);
+            return response.Content;
+        }
+
+        public Task<FileSpec> GetFileInfoAsync(string path) {
+            if (string.IsNullOrEmpty(path)) {
+                throw new ArgumentNullException(nameof(path));
+            }
+            path = NormalizePath(path);
+            try {
+                var metadata = _client.GetObjectMetadata(_bucketName, path);
+                return Task.FromResult(new FileSpec {
+                    Path = path,
+                    Size = metadata.ContentLength,
+                    Created = metadata.LastModified,
+                    Modified = metadata.LastModified
+                });
+            }
+            catch (Exception) {
+                return Task.FromResult((FileSpec)null);
+            }
+        }
+
+        public Task<bool> ExistsAsync(string path) {
+            if (string.IsNullOrEmpty(path)) {
+                throw new ArgumentNullException(nameof(path));
+            }
+            try {
+                return Task.FromResult(_client.DoesObjectExist(_bucketName, NormalizePath(path)));
+            }
+            catch (Exception ex) when (IsNotFoundException(ex)) {
+                return Task.FromResult(false);
+            }
+        }
+
+        public async Task<bool> SaveFileAsync(string path, Stream stream, CancellationToken cancellationToken = default(CancellationToken)) {
+            if (string.IsNullOrEmpty(path)) {
+                throw new ArgumentNullException(nameof(path));
+            }
+            if (!stream.CanSeek) {
+                var memory = new MemoryStream();
+                await stream.CopyToAsync(memory);
+                memory.Position = 0;
+                stream = memory;
+            }
+            try {
+                var putResult = await Task.Factory.FromAsync(
+                    (request, callback, state) => _client.BeginPutObject(request, callback, state),
+                    result => _client.EndPutObject(result), new PutObjectRequest(_bucketName, NormalizePath(path), stream), null);
+                return putResult.HttpStatusCode == HttpStatusCode.OK;

```

Comment:
Is there any docs for status codes? I know at least other api's that I'm familiar with you could get an Accepted, Ok, NotModified or 403. I feel your pains by reading their samples. They really need a complete rewrite of the whole client api using the new Task (TPL) features. It's crazy they are supporting .NET 2.0 when .NET 4.5.1 is no longer supported by Microsoft.

Reply:
Please refer to the status code docs
https://help.aliyun.com/document_detail/32005.html?spm=5176.doc31948.6.1062.2G1WCb
But I don't know how to handle it.

Reply:
Actually in China, most of the applications are built with .NET 4.5 or lower versions. and the netstandard is in estimating stage. So the alibaba do not want to support netstandard now at least. At the same time their major technology is Java, so they do not want to pay more resources for dotnet area.

Reply:
You might want to reach out to them and ask them susses status codes as they only list error status codes. If our tests are passing here then I guess it's good to go.